From ea2e269fab8d2bc3a4f7f77d091b4ae5fa2ef320 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Wed, 24 Apr 2024 16:15:00 +0200
Subject: [PATCH 0824/1695] drivers: stm32_risaf: differentiate STM32MP2x RISAF
 implementations

RISAF_REGx_CFGR and device tree encryption field lengths may change
depending on the STM32MP2x SoC.
Adapt related formulas to consider all cases.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git fa27ccf222a265969405e47d9c718c8bef83236e drivers: stm32_risaf: differentiate STM32MP2x RISAF implementations]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I5d515c3c109414012ba8ec0d6b9c49bbd19cd480
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/376413
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/firewall/stm32_risaf.c | 24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

diff --git a/core/drivers/firewall/stm32_risaf.c b/core/drivers/firewall/stm32_risaf.c
index 0f55edfdd..6cc3c70b2 100644
--- a/core/drivers/firewall/stm32_risaf.c
+++ b/core/drivers/firewall/stm32_risaf.c
@@ -82,8 +82,13 @@
 #define _RISAF_REG_CFGR_BREN		BIT(_RISAF_REG_CFGR_BREN_SHIFT)
 #define _RISAF_REG_CFGR_SEC_SHIFT	U(8)
 #define _RISAF_REG_CFGR_SEC		BIT(_RISAF_REG_CFGR_SEC_SHIFT)
+#if defined(CFG_STM32MP21)
+#define _RISAF_REG_CFGR_ENC_SHIFT	U(14)
+#define _RISAF_REG_CFGR_ENC		GENMASK_32(15, 14)
+#else /* defined(CFG_STM32MP21) */
 #define _RISAF_REG_CFGR_ENC_SHIFT	U(15)
 #define _RISAF_REG_CFGR_ENC		BIT(_RISAF_REG_CFGR_ENC_SHIFT)
+#endif /* defined(CFG_STM32MP21) */
 #define _RISAF_REG_CFGR_PRIVC_SHIFT	U(16)
 #define _RISAF_REG_CFGR_PRIVC_MASK	GENMASK_32(23, 16)
 #define _RISAF_REG_CFGR_ALL_MASK	(_RISAF_REG_CFGR_BREN | \
@@ -109,6 +114,17 @@
 #define DT_RISAF_WRITE_MASK		GENMASK_32(31, 24)
 
 #define _RISAF_GET_REGION_ID(cfg)	((cfg) & DT_RISAF_REG_ID_MASK)
+#if defined(CFG_STM32MP21)
+#define _RISAF_GET_REGION_CFG(cfg) \
+	(((((cfg) & DT_RISAF_EN_MASK) >> DT_RISAF_EN_SHIFT) << \
+	  _RISAF_REG_CFGR_BREN_SHIFT) | \
+	 ((((cfg) & DT_RISAF_SEC_MASK) >> DT_RISAF_SEC_SHIFT) << \
+	  _RISAF_REG_CFGR_SEC_SHIFT) | \
+	 ((((cfg) & DT_RISAF_ENC_MASK) >> DT_RISAF_ENC_SHIFT) << \
+	  _RISAF_REG_CFGR_ENC_SHIFT) | \
+	 ((((cfg) & DT_RISAF_PRIV_MASK) >> DT_RISAF_PRIV_SHIFT) << \
+	  _RISAF_REG_CFGR_PRIVC_SHIFT))
+#else /* defined(CFG_STM32MP21) */
 #define _RISAF_GET_REGION_CFG(cfg) \
 	(((((cfg) & DT_RISAF_EN_MASK) >> DT_RISAF_EN_SHIFT) << \
 	  _RISAF_REG_CFGR_BREN_SHIFT) | \
@@ -118,6 +134,7 @@
 	  _RISAF_REG_CFGR_ENC_SHIFT) | \
 	 ((((cfg) & DT_RISAF_PRIV_MASK) >> DT_RISAF_PRIV_SHIFT) << \
 	  _RISAF_REG_CFGR_PRIVC_SHIFT))
+#endif /* defined(CFG_STM32MP21) */
 #define _RISAF_GET_REGION_CID_CFG(cfg) \
 	(((((cfg) & DT_RISAF_WRITE_MASK) >> DT_RISAF_WRITE_SHIFT) << \
 	  _RISAF_REG_CIDCFGR_WRENC_SHIFT) | \
@@ -357,7 +374,7 @@ static TEE_Result risaf_configure_region(struct stm32_risaf_instance *risaf,
 			_RISAF_REG_CFGR_ALL_MASK,
 			cfg & _RISAF_REG_CFGR_ALL_MASK);
 
-	if ((cfg & _RISAF_REG_CFGR_ENC) == _RISAF_REG_CFGR_ENC) {
+	if (cfg & _RISAF_REG_CFGR_ENC) {
 		if (!risaf->pdata.enc_supported) {
 			EMSG("RISAF %#"PRIxPTR": encryption feature error",
 			     risaf->pdata.base.pa);
@@ -549,9 +566,14 @@ static TEE_Result stm32_risaf_pm_suspend(struct stm32_risaf_instance *risaf)
 		risaf_en = (cfg & _RISAF_REG_CFGR_BREN) << DT_RISAF_EN_SHIFT;
 		risaf_sec = ((cfg & _RISAF_REG_CFGR_SEC) >>
 			     _RISAF_REG_CFGR_SEC_SHIFT) << DT_RISAF_SEC_SHIFT;
+#if defined(CFG_STM32MP21)
+		risaf_enc = ((cfg & _RISAF_REG_CFGR_ENC) >>
+			     _RISAF_REG_CFGR_ENC_SHIFT) << DT_RISAF_ENC_SHIFT;
+#else /* defined(CFG_STM32MP21) */
 		risaf_enc = ((cfg & _RISAF_REG_CFGR_ENC) >>
 			     _RISAF_REG_CFGR_ENC_SHIFT) <<
 			    (DT_RISAF_ENC_SHIFT + 1);
+#endif /* defined(CFG_STM32MP21) */
 		risaf_priv = ((cfg & _RISAF_REG_CFGR_PRIVC_MASK) >>
 			      _RISAF_REG_CFGR_PRIVC_SHIFT) <<
 			     DT_RISAF_PRIV_SHIFT;
-- 
2.47.2

